From fce11f68491b46b93df69de0630cd9edb90bc772 Mon Sep 17 00:00:00 2001
From: Birger Koblitz <git@birger-koblitz.de>
Date: Wed, 29 Dec 2021 21:54:21 +0100
Subject: [PATCH] realtek: Create 4 different Realtek Platforms

Creates REALTEK_SOC as a basic kernel config parameter for the
RTL838X, RTL839x, RTL930X and RTL931X platforms with respective
configurations for the SoCs, which are introduced in addition.

Submitted-by: Birger Koblitz <git@birger-koblitz.de>
---
--- a/arch/mips/Kconfig
+++ b/arch/mips/Kconfig
@@ -625,27 +625,6 @@ config RALINK
 	select ARCH_HAS_RESET_CONTROLLER
 	select RESET_CONTROLLER
 
-config MACH_REALTEK_RTL
-	bool "Realtek RTL838x/RTL839x based machines"
-	select MIPS_GENERIC
-	select DMA_NONCOHERENT
-	select IRQ_MIPS_CPU
-	select CSRC_R4K
-	select CEVT_R4K
-	select SYS_HAS_CPU_MIPS32_R1
-	select SYS_HAS_CPU_MIPS32_R2
-	select SYS_SUPPORTS_BIG_ENDIAN
-	select SYS_SUPPORTS_32BIT_KERNEL
-	select SYS_SUPPORTS_MIPS16
-	select SYS_SUPPORTS_MULTITHREADING
-	select SYS_SUPPORTS_VPE_LOADER
-	select SYS_HAS_EARLY_PRINTK
-	select SYS_HAS_EARLY_PRINTK_8250
-	select USE_GENERIC_EARLY_PRINTK_8250
-	select BOOT_RAW
-	select PINCTRL
-	select USE_OF
-
 config SGI_IP22
 	bool "SGI IP22 (Indy/Indigo2)"
 	select ARC_MEMORY
@@ -1032,8 +1032,59 @@
 	  This board is based on Netlogic XLP Processor.
 	  Say Y here if you have a XLP based board.
 
+config REALTEK_SOC
+	bool "Realtek based platforms"
+	select ARCH_HAS_RESET_CONTROLLER
+	select BOOT_RAW
+	select DMA_NONCOHERENT
+	select IRQ_MIPS_CPU
+	select MIPS_GENERIC
+	select NO_EXCEPT_FILL
+	select PINCTRL
+	select RESET_CONTROLLER
+	select SYS_HAS_CPU_MIPS32_R1
+	select SYS_HAS_CPU_MIPS32_R2
+	select SYS_HAS_EARLY_PRINTK
+	select SYS_HAS_EARLY_PRINTK_8250
+	select SYS_SUPPORTS_32BIT_KERNEL
+	select SYS_SUPPORTS_BIG_ENDIAN
+	select SYS_SUPPORTS_HIGHMEM
+	select SYS_SUPPORTS_MIPS16
+	select USE_GENERIC_EARLY_PRINTK_8250
+	select USE_OF
+
 endchoice
 
+config RTL838X
+	bool "Realtek RTL838X based platforms"
+	depends on REALTEK_SOC
+	select CPU_SUPPORTS_CPUFREQ
+	select MIPS_EXTERNAL_TIMER
+
+config RTL839X
+	bool "Realtek RTL839X based platforms"
+	depends on REALTEK_SOC
+	select CPU_SUPPORTS_CPUFREQ
+	select MIPS_EXTERNAL_TIMER
+	select SYS_SUPPORTS_MULTITHREADING
+
+config RTL930X
+	bool "Realtek RTL930X based platforms"
+	depends on REALTEK_SOC
+	select MIPS_CPU_SCACHE
+	select MIPS_EXTERNAL_TIMER
+	select SYS_SUPPORTS_MULTITHREADING
+
+config RTL931X
+	bool "Realtek RTL931X based platforms"
+	depends on RTL930X
+	select MIPS_GIC
+	select COMMON_CLK
+	select CLKSRC_MIPS_GIC
+	select SYS_SUPPORTS_VPE_LOADER
+	select SYS_SUPPORTS_SMP
+	select SYS_SUPPORTS_MIPS_CPS
+
 source "arch/mips/alchemy/Kconfig"
 source "arch/mips/ath25/Kconfig"
 source "arch/mips/ath79/Kconfig"
--- a/arch/mips/generic/Makefile
+++ b/arch/mips/generic/Makefile
@@ -12,4 +12,5 @@
 obj-$(CONFIG_LEGACY_BOARD_SEAD3)	+= board-sead3.o
 obj-$(CONFIG_LEGACY_BOARD_OCELOT)	+= board-ocelot.o
 obj-$(CONFIG_MACH_INGENIC)			+= board-ingenic.o
+obj-$(CONFIG_REALTEK_SOC)		+= board-realtek.o
 obj-$(CONFIG_VIRT_BOARD_RANCHU)		+= board-ranchu.o
